knowledge-base/adr/0015-date-time-handling.md

```diff
@@ -61,3 +61,92 @@ Declined, as we are planning to release a version 2 of the SDK in the first half
+// Transformed create request
+TestEntity.requestBuilder(serializers).create(b).execute();
+```
+The middleware has a structure like below to define (de)serializers for some edm types.
+```ts
+interface DeSerializationMiddlewareInterface<T1>{
+  'Edm.String'?: {
+    serializer: (val: T1) => string;
+    deserializer: (ori: string) => T1;
+  };
+  'Edm.Int32'?: {
+     serializer: (val: T1) => string;
+     deserializer: (ori: string) => T1;
+  };
+
+}
+```
+**Please note**
+
+With the api design above, when applying it for replacing the moment lib, we will introduce multiple generic types for `Entity`, `RequestBuilder`, `TestEntity` and `TestEntityRequestBuilder`, which makes our code more complicated, but the users should not be affected.
+The number of generic types depends on the number of related edm types. 
+For example, for `ODataV4`, 4 edm types (`Edm.DateTimeOffset`, `Edm.Date`, `Edm.Duration` and `Edm.TimeOfDay`) are relevant, which means it will introduce 4 generic types for all the classes mentioned above.
+
+### Assumption
+- Request builders has the same complexity.
+- Serializer has the same complexity as the Deserializer.
+- OData V2 has the same complexity as the OData V4.
+
+### Scope
+- Type tests of the entity created by the entity builder with/without the middleware.
+```ts
+       // TestEntityTemporal<string, number>
+       const entity1 = TestEntityTemporal.builder().int32Property(1).build();
+       // TestEntityTemporal<string, string>
+       const entity2 = TestEntityTemporal.builder(customMiddlewareFull)
+         .int32Property('1')
+         .build();
+```
+- Type tests of the `GetAllRequestBuilder` with/without the middleware.
+```ts
+       // TestEntityGetAllRequestBuilder<TestEntityTemporal<string, number>, string, number>
+       const requestBuilder1 = await TestEntityTemporal.requestBuilder().getAll();
+       // TestEntityGetAllRequestBuilder<TestEntityTemporal<string, string>, string, string>
+       const requestBuilder2 = await TestEntityTemporal.requestBuilder(
+         customMiddlewareFull
+       ).getAll();
+```
+- Type tests of the return value when executing a `GetAllRequestBuilder` with/without the middleware.
+```ts
+       // TestEntityTemporal<string, number>[]
+       const res1 = await TestEntityTemporal.requestBuilder()
+         .getAll()
+         .execute(defaultDestination);
+       // TestEntityTemporal<string, number>[]
+       const res2 = await TestEntityTemporal.requestBuilder(defaultMiddleware)
+         .getAll()
+         .execute(defaultDestination);
+       // TestEntityTemporal<string, string>[]
+       const res3 = await TestEntityTemporal.requestBuilder(customMiddlewareFull)
+         .getAll()
+         .execute(defaultDestination);
+       // TestEntityTemporal<string, string>[]
+       const res4 = await TestEntityTemporal.requestBuilder(
+         customMiddlewarePartial
+       )
+         .getAll()
+         .execute(defaultDestination);
+```
```

Comment:
See comment above of giving the generic entiy TestEntityTemporal<A,B> and how it relates to the number of properties and types.